src/client/Microsoft.Identity.Client/TokenCache.ITokenCacheInternal.cs

```diff
@@ -1256,18 +1267,44 @@ bool ITokenCacheInternal.HasTokensNoLocks()
 
             foreach (MsalIdTokenCacheItem idTokenCacheItem in idTokens)
             {
-                Accessor.DeleteIdToken(idTokenCacheItem);
+                accessor.DeleteIdToken(idTokenCacheItem);
             }
 
             requestContext.Logger.Info($"Deleted {idTokens.Count} ID tokens.");
 
-            var accounts = Accessor.GetAllAccounts(partitionKey);
+            var accounts = accessor.GetAllAccounts(partitionKey);
             accounts.RemoveAll(item => !(item.HomeAccountId.Equals(account.HomeAccountId.Identifier, StringComparison.OrdinalIgnoreCase) &&
                                item.PreferredUsername.Equals(account.Username, StringComparison.OrdinalIgnoreCase)));
 
             foreach (MsalAccountCacheItem accountCacheItem in accounts)
             {
-                Accessor.DeleteAccount(accountCacheItem);
+                accessor.DeleteAccount(accountCacheItem);
+            }
+
+            if (!((ITokenCacheInternal)this).IsAppSubscribedToSerializationEvents())
+            {
+                DateTimeOffset? cacheExpiry = CalculateSuggestedCacheExpiry(accessor, requestContext.Logger);
+                await IdentityCacheWrapper.SetAsync(partitionKey, accessor, cacheExpiry).ConfigureAwait(false);
+            }
+        }
+
+        private async Task<ITokenCacheAccessor> GetOrCreateAccessorAsync(string partitionKey)
+        {
+            // If user set up legacy cache serialization, then use old accessor instance (it would have been populated with tokens)
+            // Otherwise, use IIdentityCache instance, either the user-provided or default.
+            if (((ITokenCacheInternal)this).IsAppSubscribedToSerializationEvents())
+            {
+                return Accessor;
+            }
+            else
+            {
+                var cachedAccessor = await IdentityCacheWrapper.GetAsync<ITokenCacheAccessor>(partitionKey).ConfigureAwait(false);
```

Comment:
The accessors could be simplified to just a dictionary, since we're partitioning in the cache implementation now.